ksql-streams/src/main/java/io/confluent/ksql/execution/streams/materialization/ks/KsLocator.java

```diff
@@ -51,27 +62,59 @@
   }
 
   @Override
-  @SuppressWarnings("deprecation")
-  public Optional<KsqlNode> locate(final Struct key) {
-    final StreamsMetadata metadata = kafkaStreams
-        .metadataForKey(stateStoreName, key, keySerializer);
-
-    if (metadata == StreamsMetadata.NOT_AVAILABLE) {
-      return Optional.empty();
+  public List<KsqlNode> locate(
+      final Struct key,
+      final RoutingFilter routingFilters
+  ) {
+    final KeyQueryMetadata metadata = kafkaStreams
+        .queryMetadataForKey(stateStoreName, key, keySerializer);
+
+    // Fail fast if Streams not ready. Let client handle it
+    if (metadata == KeyQueryMetadata.NOT_AVAILABLE) {
+      LOG.debug("KeyQueryMetadata not available for state store {} and key {}",
+                stateStoreName, key);
+      throw new MaterializationException(String.format(
+          "KeyQueryMetadata not available for state store %s and key %s", stateStoreName, key));
     }
 
-    final HostInfo hostInfo = metadata.hostInfo();
-    return Optional.of(asNode(hostInfo));
+    final HostInfo activeHost = metadata.getActiveHost();
+    final Set<HostInfo> standByHosts = metadata.getStandbyHosts();
+    LOG.info("Before filtering: Active host {} , standby hosts {}", activeHost, standByHosts);
+
+    final Stream<KsqlHost> active = Stream.of(asKsqlHost(activeHost));
+    final Stream<KsqlHost> standby = standByHosts
+        .stream()
+        .map(this::asKsqlHost);
+    final Stream<KsqlHost> hostStream = Stream.concat(active, standby);
+
+    // Filter out hosts based on liveness and lag filters.
+    // The list is ordered by routing preference: active node is first, then standby nodes
+    // in order of increasing lag.
+    // If heartbeat is not enabled, all hosts are considered alive.
+    final List<KsqlNode> filteredHosts = hostStream
+        .filter(hostInfo -> routingFilters.filter(
+              activeHost, hostInfo, stateStoreName, metadata.getPartition()))
+        .map(this::asNode)
+        .collect(Collectors.toList());
+
+    LOG.info("Filtered and ordered hosts: {}", filteredHosts);
```

Comment:
lets also move this to DEBUG or TRACE.. can you please make a pass and remove any INFO level log statements from the pull query execution path